### PR TITLE
chore(auth): add oauth metadata into token orchestrator

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

This change adds oauth metadata into the token orchestrator. This metadata is set after the user has authenticated via OAuth, and it can be accessed from the current `KeyValueStorage` (LocalStorage, SessionStorage, CookieStorage, etc).

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.ts [hunk 1]
```diff
@@ -5,21 +5,24 @@ import { CognitoUserPoolConfig } from '@aws-amplify/core';
...
 ): Promise<void | OpenAuthSessionResult> => {
 	const { isOAuthSignIn } = await store.loadOAuthSignIn();
+	const oauthMetadata = await tokenOrchestrator.getOAuthMetadata();
 
 	// Clear everything before attempting to visted logout endpoint since the current application
 	// state could be wiped away on redirect
 	await completeOAuthSignOut(store);
 
-	if (isOAuthSignIn) {
+	if (isOAuthSignIn || oauthMetadata?.oauthSignIn) {
```
Comment (AllanZhengYP) [inline packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.ts:25]: Nit: some inline doc on why checking both places is helpful here.

**Review by AllanZhengYP — APPROVED** [review]

File: packages/auth/src/providers/cognito/utils/oauth/completeOAuthFlow.ts [hunk 2]
```diff
@@ -148,6 +149,9 @@ const handleCodeFlow = async ({
 		TokenType: token_type,
 		ExpiresIn: expires_in,
 	});
+	await tokenOrchestrator.setOAuthMetadata({
```
Comment (HuiSF) [inline packages/auth/src/providers/cognito/utils/oauth/completeOAuthFlow.ts:152]: Can probably relocate this to the `completeFlow` function, so we don't need to do this in both flows respectively? Wdyt?

File: packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.ts (hunk 1 above)
Comment (jimblanc) [reply packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.ts:25]: +1, at first glance `isOAuthSignIn` & `oauthSignIn` seem like they'd have very similar purposes

**Review by jimblanc — APPROVED** [review]

File: packages/auth/__tests__/providers/cognito/utils/oauth/handleOAuthSignOut.test.ts [hunk 3]
```diff
@@ -39,7 +45,21 @@ describe('handleOAuthSignOut', () => {
...
 			preferPrivateSession: false,
 		});
-		await handleOAuthSignOut(cognitoConfig, mockStore);
+		await handleOAuthSignOut(cognitoConfig, mockStore, mockTokenOrchestrator);
+
+		expect(mockCompleteOAuthSignOut).toHaveBeenCalledWith(mockStore);
+		expect(mockOAuthSignOutRedirect).toHaveBeenCalledWith(cognitoConfig);
+	});
+
+	it('should complete OAuth sign out and redirect when there oauth metadata in tokenOrchestrator', async () => {
```
Comment (jimblanc) [inline packages/auth/__tests__/providers/cognito/utils/oauth/handleOAuthSignOut.test.ts:54]: Are there more tests we can add? E.g. signing out of an existing session without this meta-data in storage

Comment (israx) [reply packages/auth/__tests__/providers/cognito/utils/oauth/handleOAuthSignOut.test.ts:54]: The previous tests are already signing out without depending on this metadata in storage already. Further test cases will be introduced by the e2e test

File: packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.ts (hunk 1 above)
Comment (HuiSF) [reply packages/auth/src/providers/cognito/utils/oauth/handleOAuthSignOut.ts:31]: Thanks for adding the comments, eventually the localStorage based oAuthStore can be remove, is that correct? If so, worth to add a TODO comment?

**Review by HuiSF — APPROVED** [review]

**Pull request merged**